### PR TITLE
[Feature](nereids) support sub-query and alias in FromClause

### PR DESCRIPTION
**Pull request by sohardforaname**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Support sub-query and alias for TPC-H，for example:
select * from (select * from (T1) A join T2 as B on T1.id = T2.id) T;

## Checklist(Required)

1. Does it affect the original behavior: (Yes)
2. Has unit tests been added: (Yes)
3. Has document been added or modified: (No)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java [hunk 1]
```diff
@@ -118,6 +121,38 @@ private Plan groupToTreeNode(Group group) {
...
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            GroupExpressionAdapter that = (GroupExpressionAdapter) o;
+            if (that.groupExpr.getPlan() instanceof LogicalOlapScan) {
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java:144]: why special judgment for `OlapScan`?

**Comment by jackwener** [comment]
I recommend that you can learn use ` git rebase `.
Use `git pull upstream master --rebase`.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java [hunk 2]
```diff
@@ -51,64 +51,64 @@ public class BindSlotReference implements AnalysisRuleFactory {
...
-            ),
-            RuleType.BINDING_SORT_SLOT.build(
-                logicalSort().then(sort -> {
-                    List<OrderKey> sortItemList = sort.getOrderKeys()
-                            .stream()
-                            .map(orderKey -> {
-                                Expression item = bind(orderKey.getExpr(), sort.children(), sort);
-                                return new OrderKey(item, orderKey.isAsc(), orderKey.isNullFirst());
-                            }).collect(Collectors.toList());
+                RuleType.BINDING_PROJECT_SLOT.build(
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java:54]: Reduce the `code-style-change`.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java (hunk 1 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java:144]: The class is a package class for GroupExpression because the equals() of the latter can not divide UnboundRelation from LogicalScanOlap as their logical properties are the same, but actually they are not the same.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java (hunk 2 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java:54]: ok, thx

File: fe/fe-core/pom.xml [hunk 3]
```diff
@@ -668,6 +668,12 @@ under the License.
             <artifactId>maven-compiler-plugin</artifactId>
             <version>3.10.1</version>
         </dependency>
+<!--        <dependency>-->
```
Comment (morrySnow) [inline fe/fe-core/pom.xml:671]: revert these changes

File: docs/zh-CN/developer/developer-guide/fe-idea-dev.md [hunk 4]
```diff
@@ -32,6 +32,15 @@ under the License.
 
 安装 JDK1.8+ ，使用 IntelliJ IDEA 打开 FE.
 
+#### MacOS下载
```
Comment (morrySnow) [inline docs/zh-CN/developer/developer-guide/fe-idea-dev.md:35]: revert these changes

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisLexer.tokens [hunk 5]
```diff
@@ -0,0 +1,620 @@
+SEMICOLON=1
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisLexer.tokens:1]: this file should not add to source

File: fe/fe-core/src/main/java/org/apache/doris/analysis/SchemaTableType.java [hunk 6]
```diff
@@ -68,6 +68,7 @@ public enum SchemaTableType {
     SCH_INVALID("NULL", "NULL", TSchemaTableType.SCH_INVALID);
     private static final String dbName = "INFORMATION_SCHEMA";
     private static SelectList fullSelectLists;
+
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/SchemaTableType.java:71]: revert these changes

File: fe/fe-core/src/main/java/org/apache/doris/catalog/PrimitiveType.java [hunk 7]
```diff
@@ -92,6 +92,7 @@ public enum PrimitiveType {
     }
 
     private static ImmutableSetMultimap<PrimitiveType, PrimitiveType> implicitCastMap;
+
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/PrimitiveType.java:95]: revert these changes

File: fe/fe-core/src/main/java/org/apache/doris/catalog/View.java [hunk 8]
```diff
@@ -45,7 +45,7 @@
  * Table metadata representing a catalog view or a local view from a WITH clause.
  * Most methods inherited from Table are not supposed to be called on this class because
  * views are substituted with their underlying definition during analysis of a statement.
- *
+ * <p>
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/View.java:48]: revert these changes

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlCommand.java [hunk 9]
```diff
@@ -51,6 +51,7 @@ public enum MysqlCommand {
     COM_RESET_CONNECTION("COM_RESET_CONNECTION", 31);
 
     private static Map<Integer, MysqlCommand> codeMap = Maps.newHashMap();
+
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlCommand.java:54]: revert these changes

File: fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java [hunk 10]
```diff
@@ -124,6 +125,10 @@ private void analyze() {
         new AnalyzeRulesJob(plannerContext).execute();
     }
 
+    private void finalizeAnalyze() {
+
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java:129]: still empty?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java [hunk 11]
```diff
@@ -131,11 +172,16 @@ private Plan groupToTreeNode(Group group) {
      */
     private GroupExpression insertOrRewriteGroupExpression(GroupExpression groupExpression, Group target,
             boolean rewrite, LogicalProperties logicalProperties) {
-        GroupExpression existedGroupExpression = groupExpressions.get(groupExpression);
-        if (existedGroupExpression != null) {
+        GroupExpressionAdapter adapter = new GroupExpressionAdapter(groupExpression);
+        // GroupExpression existedGroupExpression = groupExpressions.get(groupExpression);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java:176]: remove useless code directly

File: fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+        for (Expression child : children) {
+            if (!checkExpressionBound(child)) {
+                return false;
+            }
+        }
+        return true;
+    }
+}
+
+/*
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java:241]: remove these

**Review by morrySnow — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Slot.java [hunk 13]
```diff
@@ -30,4 +32,8 @@ public Slot toSlot() {
     public Slot withNullable(boolean newNullable) {
         throw new RuntimeException("Do not implement");
     }
+
+    public Slot withQualifier(List<String> qualifiers) {
+        throw new RuntimeException("Do not implement");
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Slot.java:37]: "Not implemented" may be a better exception message here

Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Slot.java:37]: ok，I will change it.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalSubQueryAlias.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+    @Override
+    public List<Expression> getExpressions() {
+        return Collections.emptyList();
+    }
+
+    @Override
+    public Plan withGroupExpression(Optional<GroupExpression> groupExpression) {
+        return new LogicalSubQueryAlias<>(alias, groupExpression, Optional.of(logicalProperties), child());
+    }
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalSubQueryAlias.java:107]: Will `LogicalSubQueryAlias`  still exist in the Memo? If not, I think it'd be better to not overwrite this method.

Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalSubQueryAlias.java:107]: Will not, but it's not need to throw a exception as their is always a node below the node so the child() will never be null

Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalSubQueryAlias.java:107]: Ok, thx

File: fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java [hunk 15]
```diff
@@ -104,13 +105,18 @@ public PhysicalPlan plan(LogicalPlan plan, PhysicalProperties outputProperties,
...
+        finalizeAnalyze();
         rewrite();
         optimize();
 
         // Get plan directly. Just for SSB.
         return getRoot().extractPlan();
     }
 
+    private void finalizeAnalyze() {
+        new FinalizeAnalyzeJob(plannerContext).execute();
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java:117]: Do we really need a specific job to do this?

Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java:117]: yes, because we should eliminate all the logical-subquery-alias nodes, or the rewrite phase will core.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java [hunk 16]
```diff
@@ -167,6 +168,9 @@ public boolean equals(Object o) {
             return false;
         }
         GroupExpression that = (GroupExpression) o;
+        if (plan instanceof LogicalOlapScan) {
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java:171]: You should add a comment here to show why LogicalOlapScan is not equals

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 17]
```diff
@@ -190,11 +194,38 @@ public LogicalPlan visitRegularQuerySpecification(RegularQuerySpecificationConte
     /**
      * Create an aliased table reference. This is typically used in FROM clauses.
      */
+    private LogicalPlan applyAlias(TableAliasContext ctx, LogicalPlan plan) {
+        if (null != ctx.strictIdentifier()) {
+            String alias = ctx.strictIdentifier().getText();
+            if (null != ctx.identifierList()) {
+                // List<String> colName = visitIdentifierSeq(ctx.identifierList().identifierSeq());
+                // TODO: multi-colName
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:202]: You should throw a ParseException here

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java [hunk 18]
```diff
@@ -33,6 +33,7 @@ public enum RuleType {
     BINDING_SORT_SLOT(RuleTypeClass.REWRITE),
     BINDING_PROJECT_FUNCTION(RuleTypeClass.REWRITE),
     BINDING_AGGREGATE_FUNCTION(RuleTypeClass.REWRITE),
+    BINDING_ALIAS_SLOT(RuleTypeClass.REWRITE),
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java:36]:
```suggestion
    BINDING_ALIAS_SUBQUERY_SLOT(RuleTypeClass.REWRITE),
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSubQueryAlias.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Rule to bind relation when encounter sub query and alias
+ */
+public class BindSubQueryAlias implements AnalysisRuleFactory {
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.of(
+                RuleType.BINDING_ALIAS_SLOT.build(
+                        logicalSubQueryAlias().then(alias -> new LogicalSubQueryAlias<>(alias.getAlias(), alias.child())
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSubQueryAlias.java:36]: this line do nothing?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+ * If we match the alias node and return its child node, in the execute() of the job
+ *
+ * TODO: refactor group merge strategy to support the feature above
+ */
+public class EliminateAliasNode implements AnalysisRuleFactory {
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.of(
+                RuleType.PROJECT_ELIMINATE_ALIAS_NODE.build(
+                        logicalProject().then(project -> eliminateSubQueryAliasNode(project, project.children()))
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java:44]: why not use this pattern and save work to judge child type is LogicalSubqueryAlias?

```
logicalProject(logicalSubqueryAlias()).then(...)
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java (hunk 16 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java:171]: Ok，I add it now.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSubQueryAlias.java (hunk 19 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSubQueryAlias.java:36]: No, if delete that, it will core when analyze.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java (hunk 20 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java:44]: That is really ok！Thx.

File: fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+    }
+
+    private void checkAnalyze(String sql) {
+        LogicalPlan analyzed = analyze(sql);
+        System.out.println(analyzed.treeString());
+        Assertions.assertTrue(checkBound(analyzed));
+    }
+
+    private void finalizeAnalyze(String sql) {
+        Memo memo = new Memo(parser.parseSingle(sql));
```
Comment (wangshuo128) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java:162]: Please use `PlanRewriter` to simplify the code.

File: fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+            LogicalPlan parsed = parser.parseSingle(sql);
+            System.out.println(parsed.treeString());
+            return analyze(parsed, connectContext);
+        } catch (Throwable t) {
+            throw new IllegalStateException("Analyze failed", t);
+        }
+    }
+
+    private LogicalPlan analyze(LogicalPlan inputPlan, ConnectContext connectContext) {
+        Memo memo = new Memo(inputPlan);
```
Comment (wangshuo128) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java:189]: ditto.

File: fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+                new BindFunction(),
+                new BindRelation(),
+                new BindSubQueryAlias(),
+                new BindSlotReference(),
+                new ProjectToGlobalAggregate());
+        return (LogicalPlan) memo.copyOut();
+    }
+
+    private void executeRewriteBottomUpJob(PlannerContext plannerContext, RuleFactory... ruleFactory) {
+        Group rootGroup = plannerContext.getMemo().getRoot();
```
Comment (wangshuo128) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java:205]: ditto.

File: fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+                new BindRelation(),
+                new BindSubQueryAlias(),
+                new BindSlotReference(),
+                new ProjectToGlobalAggregate());
+        System.out.println(memo.copyOut().treeString());
+        new FinalizeAnalyzeJob(plannerContext).execute();
+        System.out.println(memo.copyOut().treeString());
+    }
+
+    private LogicalPlan analyze(String sql) {
```
Comment (wangshuo128) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java:178]: Please use `NereidsAnalyzer` in the latest code.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java (hunk 20 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java:44]: > why not use this pattern and save work to judge child type is LogicalSubqueryAlias?
> 
> ```java
> logicalProject(logicalSubqueryAlias()).then(...)
> ```
logicalJoin cannot use the pattern, others are ok

Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java]: i think logicalJoin could use
```
logicalJoin(logicalSubqueryAlias(), group())

+

logicalJoin(group(), logicalSubqueryAlias()) 
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSubQueryAlias.java (hunk 19 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSubQueryAlias.java]: just replace logical properties implicitly

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java [hunk 25]
```diff
@@ -168,6 +169,11 @@ public boolean equals(Object o) {
             return false;
         }
         GroupExpression that = (GroupExpression) o;
+        // if the plan is LogicalOlapScan, this == that should be true,
+        // because equals() can not divide UnboundRelation and the plan above.
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java:173]: this comment is not clear to describe the question. we need to explain that if one relation appear in plan more than once, we cannot distinguish them throw equals function, since equals function cannot use output info.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 26]
```diff
@@ -190,11 +194,39 @@ public LogicalPlan visitRegularQuerySpecification(RegularQuerySpecificationConte
     /**
      * Create an aliased table reference. This is typically used in FROM clauses.
      */
+    private LogicalPlan applyAlias(TableAliasContext ctx, LogicalPlan plan) {
+        if (null != ctx.strictIdentifier()) {
+            String alias = ctx.strictIdentifier().getText();
+            if (null != ctx.identifierList()) {
+                throw new IllegalStateException("Do not implemented");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:201]: use ParseException instead of IllegalStateException

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 27]
```diff
@@ -190,11 +194,39 @@ public LogicalPlan visitRegularQuerySpecification(RegularQuerySpecificationConte
     /**
      * Create an aliased table reference. This is typically used in FROM clauses.
      */
+    private LogicalPlan applyAlias(TableAliasContext ctx, LogicalPlan plan) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:197]:
```suggestion
    private LogicalPlan withTableAlias(LogicalPlan plan, TableAliasContext ctx) {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 28]
```diff
@@ -190,11 +194,39 @@ public LogicalPlan visitRegularQuerySpecification(RegularQuerySpecificationConte
...
-        // TODO: sample and time travel, alias, sub query
-        return new UnboundRelation(tableId);
+        return applyAlias(ctx.tableAlias(), new UnboundRelation(tableId));
+    }
+
+    @Override
+    public LogicalPlan visitAliasedQuery(AliasedQueryContext ctx) {
+        String alias = "__auto_generated_name__";
+        LogicalPlan query = visitQuery(ctx.query());
+        if (null != ctx.tableAlias().strictIdentifier()) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:221]: strictIdentifier cannot be null

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+                                .getLogicalExpression()
+                                .getPlan();
+                    }
+                    return child;
+                })
+                .collect(Collectors.toList());
+        return (LogicalPlan) node.withChildren(nodes);
+    }
+
+    private boolean checkIsSubQueryAliasNode(Plan node) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java:83]: why not use GroupPlan as parameter's type directly?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java (hunk 28 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:221]: Yes, I got an error at running query.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java (hunk 29 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java:83]: Yes, really Ok hhh

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java [hunk 30]
```diff
@@ -168,6 +169,11 @@ public boolean equals(Object o) {
...
+        if (plan instanceof LogicalOlapScan) {
```
[the remark this answers is not in the document]
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java [hunk 31]
```diff
@@ -168,6 +169,12 @@ public boolean equals(Object o) {
             return false;
         }
         GroupExpression that = (GroupExpression) o;
+        // if the plan is LogicalRelation, this == that should be true,
+        // when if one relation appear in plan more than once,
+        // we cannot distinguish them throw equals function, since equals function cannot use output info.
+        if (plan instanceof LogicalRelation) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java:175]: should also add physical relation

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 32]
```diff
@@ -206,11 +211,33 @@ public LogicalPlan visitRegularQuerySpecification(RegularQuerySpecificationConte
     /**
      * Create an aliased table reference. This is typically used in FROM clauses.
      */
+    private LogicalPlan withTableAlias(LogicalPlan plan, TableAliasContext ctx) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:214]: add a Developing annotation

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java (hunk 20 above)
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java]: or you can use
```java
logicalJoin(any(), any())
  .when(join -> join.left() instanceof LogicalSubqueryAlias || join.right() instanceof LogicalSubqueryAlias)
  .then(join -> ...)
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java [hunk 33]
```diff
@@ -0,0 +1,91 @@
...
+        return (LogicalPlan) node.withChildren(nodes);
+    }
+
+    private boolean checkIsSubQueryAliasNode(GroupPlan node) {
+        return node.getGroup().getLogicalExpression().getPlan().getType()
+                == PlanType.LOGICAL_SUBQUERY_ALIAS;
+    }
+
+    private Plan getPlan(Plan node) {
+        return ((GroupPlan) node.child(0)).getGroup().getLogicalExpression().getPlan();
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java:89]: get children plan from GroupPlan is not recommended.

I think you can simple return LogicalSubQueryAlias's children plan which is groupPlan.

```java
logicalProject(logicalSubQueryAlias()).then(project -> {
  return project.withChildren(project.child().child());
});

logicalJoin(logicalSubQueryAlias(), groupPlan()).then(join -> {
  return join.withChildren(ImmutableList.of(join.left().child(), join.right()));
});

logicalJoin(groupPlan(), logicalSubQueryAlias()).then(join -> {
  return join.withChildren(ImmutableList.of(join.left(), join.right().child()));
});
```

@morrySnow do you think so?

Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java:89]: I've tried this, but unluckily it will core at Select * From T1 T;

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java (hunk 32 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:214]: ok

File: fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java (hunk 31 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/memo/GroupExpression.java:175]: ok！

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java (hunk 33 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java:89]: It works, I've rewritten the code, thx.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+ * TODO: refactor group merge strategy to support the feature above
+ */
+public class EliminateAliasNode implements AnalysisRuleFactory {
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.of(
+                RuleType.PROJECT_ELIMINATE_ALIAS_NODE.build(
+                        logicalProject(logicalSubQueryAlias())
+                                .then(project -> project.withChildren(ImmutableList.of(project.child().child())))
+                //.then(project -> eliminateSubQueryAliasNode(project, project.children()))
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java:45]: please remove useless comment and useless code~

Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/EliminateAliasNode.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 35]
```diff
@@ -207,11 +213,33 @@ public LogicalPlan visitRegularQuerySpecification(RegularQuerySpecificationConte
...
+    }
+
+    @Override
+    public LogicalPlan visitAliasedQuery(AliasedQueryContext ctx) {
+        return withTableAlias(visitQuery(ctx.query()), ctx.tableAlias());
+    }
+
+    @Override
+    public LogicalPlan visitAliasedRelation(AliasedRelationContext ctx) {
+        return withTableAlias((LogicalPlan) visitRelation(ctx.relation()), ctx.tableAlias());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:242]: abstract visitRelation from visitFromClause

Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:242]: what does abstract mean? change visitRelation to visitFromClause?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java [hunk 36]
```diff
@@ -69,4 +68,33 @@ private Table getTable(String dbName, String tableName, Env env) {
...
+            dbName = ctx.getClusterName() + ":" + nameParts.get(0);
+        }
+        Table table = getTable(dbName, nameParts.get(1), ctx.getEnv());
+        if (table.getType() == TableType.OLAP) {
+            return new LogicalOlapScan(table, ImmutableList.of(dbName));
+        } else if (table.getType() == TableType.VIEW) {
+            LogicalPlan viewPlan = new NereidsAnalyzer(ctx).analyze(table.getDdlSql());
+            return new LogicalSubQueryAlias<>(table.getName(), viewPlan);
+        }
+        throw new RuntimeException("");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java:98]: need to add some meaningful message in exception

File: regression-test/conf/regression-conf.groovy [hunk 37]
```diff
@@ -20,11 +20,11 @@
 // **Note**: default db will be create if not exist
 defaultDb = "regression_test"
 
-jdbcUrl = "jdbc:mysql://127.0.0.1:9030/?"
+jdbcUrl = "jdbc:mysql://127.0.0.1:6786/?"
```
Comment (morrySnow) [inline regression-test/conf/regression-conf.groovy:23]: change back to original port

File: regression-test/conf/regression-conf.groovy [hunk 38]
```diff
@@ -20,11 +20,11 @@
...
 jdbcUser = "root"
 jdbcPassword = ""
 
-feHttpAddress = "127.0.0.1:8030"
+feHttpAddress = "127.0.0.1:6780"
```
Comment (morrySnow) [inline regression-test/conf/regression-conf.groovy:27]: ditto

File: regression-test/suites/nereids_syntax_p0/load.groovy [hunk 39]
```diff
@@ -181,4 +181,6 @@ suite("load") {
         (1309892, 1, 1303, 1432, 15, 19920517, "3-MEDIUM", 0, 24, 2959704, 5119906, 7, 2752524, 73992, 0, 19920619, "TRUCK"),
         (1310179, 6, 1312, 1455, 29, 19921110, "3-MEDIUM", 0, 15, 1705830, 20506457, 10, 1535247, 68233, 8, 19930114, "FOB");
     """
+
+    sleep(6000)
```
Comment (morrySnow) [inline regression-test/suites/nereids_syntax_p0/load.groovy:185]: why need to sleep?

File: fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java [hunk 40]
```diff
@@ -0,0 +1,196 @@
...
+        LogicalPlan analyzed = analyze(sql);
+        System.out.println(analyzed.treeString());
+        Assertions.assertTrue(checkBound(analyzed));
+    }
+
+    private void finalizeAnalyze(String sql) {
+        LogicalPlan plan = analyze(sql);
+        System.out.println(plan.treeString());
+        plan = (LogicalPlan) PlanRewriter.bottomUpRewrite(plan, connectContext, new EliminateAliasNode());
+        System.out.println(plan.treeString());
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java:156]: remove useless print

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 41]
```diff
@@ -211,11 +217,33 @@ public LogicalPlan visitRegularQuerySpecification(RegularQuerySpecificationConte
...
```
[the remark this answers is not in the document]
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:246]: visitRelation return Object, but withAlias require LogicalPlan

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 42]
```diff
@@ -567,23 +595,45 @@ public List<Expression> visitNamedExpressionSeq(NamedExpressionSeqContext namedC
...
 
+    /**
+     * Create OrderKey list.
+     *
+     * @param ctx QueryOrganizationContext
+     * @return List of OrderKey
+     */
     @Override
-    public LogicalPlan visitFromClause(FromClauseContext ctx) {
+    public List<OrderKey> visitQueryOrganization(QueryOrganizationContext ctx) {
```
[the remark this answers is not in the document]
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:605]: done

File: fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java [hunk 43]
```diff
@@ -0,0 +1,194 @@
...
+                    PhysicalProperties.ANY,
+                    connectContext
+            );
+            // Just to check whether translate will throw exception
+            new PhysicalPlanTranslator().translatePlan(plan, new PlanTranslatorContext());
+        }
+    }
+
+    @Test
+    public void testCase1() {
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/util/AnalyzeSubQueryTest.java:97]: give test case a meaningful name to explain what it is test

**Review by morrySnow — DISMISSED** [review]

**Review by morrySnow — APPROVED** [review]

**Review by 924060929 — APPROVED** [review]
LGTM
